Setup publishing of multiplatform dummy library.

pocmo opened the pull request:
This adds the gradle config for publishing the multiplatform library. Luckily the multiplatform does most of the heavy lifting. I expect this to not work on taskcluster though: Other than the other components, this one will create multiple artifacts instead of just one.

This is how it looks like if, for testing, I publish this to my GitHub packages repository:
https://github.com/pocmo/android-components/packages

pocmo commented:
This is what the component publishes when publishing to a local maven repository:

```
.
├── multiplatform-lib-dummy
│   ├── 71.0.0
│   │   ├── multiplatform-lib-dummy-71.0.0.jar
│   │   ├── multiplatform-lib-dummy-71.0.0.module
│   │   └── multiplatform-lib-dummy-71.0.0.pom
│   └── maven-metadata-local.xml
├── multiplatform-lib-dummy-android
│   ├── 71.0.0
│   │   ├── multiplatform-lib-dummy-android-71.0.0-sources.jar
│   │   ├── multiplatform-lib-dummy-android-71.0.0.aar
│   │   ├── multiplatform-lib-dummy-android-71.0.0.module
│   │   └── multiplatform-lib-dummy-android-71.0.0.pom
│   └── maven-metadata-local.xml
├── multiplatform-lib-dummy-iosarm64
│   ├── 71.0.0
│   │   ├── multiplatform-lib-dummy-iosarm64-71.0.0-sources.jar
│   │   ├── multiplatform-lib-dummy-iosarm64-71.0.0.klib
│   │   ├── multiplatform-lib-dummy-iosarm64-71.0.0.module
│   │   └── multiplatform-lib-dummy-iosarm64-71.0.0.pom
│   └── maven-metadata-local.xml
├── multiplatform-lib-dummy-iosx64
│   ├── 71.0.0
│   │   ├── multiplatform-lib-dummy-iosx64-71.0.0-sources.jar
│   │   ├── multiplatform-lib-dummy-iosx64-71.0.0.klib
│   │   ├── multiplatform-lib-dummy-iosx64-71.0.0.module
│   │   └── multiplatform-lib-dummy-iosx64-71.0.0.pom
│   └── maven-metadata-local.xml
└── multiplatform-lib-dummy-metadata
    ├── 71.0.0
    │   ├── multiplatform-lib-dummy-metadata-71.0.0-sources.jar
    │   ├── multiplatform-lib-dummy-metadata-71.0.0.jar
    │   ├── multiplatform-lib-dummy-metadata-71.0.0.module
    │   └── multiplatform-lib-dummy-metadata-71.0.0.pom
    └── maven-metadata-local.xml
```

pocmo commented:
@MihaiTabara @escapewindow I wonder what the best way is to get something like this published on `(nightly.)maven.mozilla.org`? I assume just landing this as-is will fail since all other components just create a single artifact? And this one creates `aar`, `jar`, `klib` and `module` artifacts. 🤔 What would be the best next steps? Or is this a big project to get setup?

escapewindow commented:
@pocmo I think we need to answer:
- what artifacts do we want to upload?
- should they all get uploaded to the same location as existing artifacts, or do we need a different path?
- does the current build artifacts logic need to special case the multiplatform component?

It looks like beetmover [takes its artifact list from its upstream deps](https://github.com/mozilla-mobile/android-components/blob/master/taskcluster/ac_taskgraph/transforms/beetmover.py#L40-L58). These will get uploaded to [this path](https://github.com/mozilla-mobile/android-components/blob/master/taskcluster/ci/beetmover/kind.yml#L29) in the bucket.

I took a look at [treeherder](https://treeherder.mozilla.org/jobs?repo=android-components) since it looks like signing+beetmover only run on nightly and release graphs.
- [build-nightly-feature-downloads](https://firefox-ci-tc.services.mozilla.com/tasks/aNxVn3hQTvOhJOgJuoKMgA)
- [signing-nightly-feature-downloads](https://firefox-ci-tc.services.mozilla.com/tasks/InGE3QHpQVOB4xhwFgszHw)
- [beetmover-nightly-feature-downloads](https://firefox-ci-tc.services.mozilla.com/tasks/Dk2NJXQLTEC0n90e5uVX4A)

The beetmover task's `upstreamArtifacts` appears to take some files from the build task and some files from the signing task. The build task exposes an artifacts dictionary in its payload; we use this dictionary to craft the signing and beetmover tasks. So depending on the answers to the above questions, this may be as simple as just populating the artifacts dictionary properly in the build task, if we want the exact same behavior for all artifacts. If we want a different location for iOS vs android artifacts, we need to expose more information in the dictionary, and alter where we upload the artifacts on the beetmover side.

rocketsroger commented:
My mistake.  Looks like there are some questions.

pocmo commented:
I _think_ we would want to publish all of them. But now that you ask, I think I want to look more at the integration side with Jeff before pushing this forward.

pocmo commented:
@escapewindow So, we looked into this..

> what artifacts do we want to upload?

We would like all of those artifacts uploaded to our maven repository. The main difference to other components is that we have multiple artifacts (~5) instead of just one, and different outputs: aar (that's what our other Android components output too), jar (we have one other component that does this, and for that we added `artifact-type: jar` to `.buildconfig.yml`) and `klib` (new). In addition to that they all include this `module` file that we may need too.

> should they all get uploaded to the same location as existing artifacts, or do we need a different path?

I am not exactly sure what you mean with path in this case. The path they will end up on `maven.mozilla.org`? I think this is flexible. In the publishing code I changed the group ID to be `org.mozilla.multiplatform` instead of `org.mozilla.components` to separate them.. but we can change that to whatever makes sense.

escapewindow commented:
> @escapewindow So, we looked into this..
> 
> > what artifacts do we want to upload?
> 
> We would like all of those artifacts uploaded to our maven repository. The main difference to other components is that we have multiple artifacts (~5) instead of just one, and different outputs: aar (that's what our other Android components output too), jar (we have one other component that does this, and for that we added `artifact-type: jar` to `.buildconfig.yml`) and `klib` (new). In addition to that they all include this `module` file that we may need too.

Ok. And are those showing up in the build task's artifact dictionary, or no?

> > should they all get uploaded to the same location as existing artifacts, or do we need a different path?
> 
> I am not exactly sure what you mean with path in this case. The path they will end up on `maven.mozilla.org`? I think this is flexible. In the publishing code I changed the group ID to be `org.mozilla.multiplatform` instead of `org.mozilla.components` to separate them.. but we can change that to whatever makes sense.

Yes. Currently the beetmover task hardcodes this to [`maven2/org/mozilla/components/{component}/{version}/{artifact_file_name}`](https://github.com/mozilla-mobile/android-components/blob/master/taskcluster/ci/beetmover/kind.yml#L29) Will that work? Or do we want to specify a different path template for some of the artifacts?

pocmo commented:
> Ok. And are those showing up in the build task's artifact dictionary, or no?

I am actually not sure. I'll have to read the automation code. Is there an easy way to run this and look at the graph without landing this and blowing up everything? :)

> Yes. Currently the beetmover task hardcodes this to [`maven2/org/mozilla/components/{component}/{version}/{artifact_file_name}`](https://github.com/mozilla-mobile/android-components/blob/master/taskcluster/ci/beetmover/kind.yml#L29) Will that work? Or do we want to specify a different path template for some of the artifacts?

For `org/mozilla/components`, I think we could use that, I picked `org/mozilla/multiplatform` to separate the components, but there's no need for that, and we can use `org/mozilla/components` if that is easier.

For `{component}/{version}/{artifact_file_name}` this seems more complicated now. In the example here, we have only one (gradle) component `multiplatform-lib-dummy`, but output the tree from https://github.com/mozilla-mobile/android-components/pull/9435#issuecomment-762800776 - So kind of a 1:n mapping instead of a 1:1 mapping as before? 🤔

escapewindow commented:
> > Ok. And are those showing up in the build task's artifact dictionary, or no?
> 
> I am actually not sure. I'll have to read the automation code. Is there an easy way to run this and look at the graph without landing this and blowing up everything? :)

Yes!

First, install taskgraph in a virtualenv. Some docs are [here](https://docs.mozilla-releng.net/en/latest/procedures/mobile_automation_setup.html?highlight=mobile#how-to-set-up-taskgraph-for-mobile).

Second, find an appropriate `parameters.yml` file. These will have different values for a PR, versus a master or main merge or push, versus a release tag, versus a cron task, versus an action (including the release-promotion action that shipit uses). If, for instance, you wanted to test an on-master-merge graph, then you could find the decision task that ran against the latest master merge commit, and download `parameters.yml` from its artifacts.

`taskgraph target-graph -p PARAMETERS_FILE --json | tee graph.json` will then generate the entire graph that would be generated on-master-merge, using the local clone of android-components or fenix that you're cd'ed into.

Even better, you can create a `before.json` with an unpatched source tree, and an `after.json` from your patched source tree, which you can then diff to see what exactly changed between the two runs.

And even better still, we've gathered example parameters files for later usage and diffing, e.g. [these for android-components](https://hg.mozilla.org/build/braindump/file/tip/taskcluster/taskgraph-diff/params-android-components). Our [`taskgraph-gen`](https://hg.mozilla.org/build/braindump/file/tip/taskcluster/taskgraph-diff/taskgraph-gen.py) tool creates a task graph per parameters file using the current source; our [`taskgraph-diff`](https://hg.mozilla.org/build/braindump/file/tip/taskcluster/taskgraph-diff/taskgraph-diff.py) tool takes 2 sets of those task graph files and diffs them for inspection. Docs for that are also [here](https://docs.mozilla-releng.net/en/latest/procedures/mobile_automation_setup.html?highlight=mobile#how-to-set-up-taskgraph-for-mobile).


> > Yes. Currently the beetmover task hardcodes this to [`maven2/org/mozilla/components/{component}/{version}/{artifact_file_name}`](https://github.com/mozilla-mobile/android-components/blob/master/taskcluster/ci/beetmover/kind.yml#L29) Will that work? Or do we want to specify a different path template for some of the artifacts?
> 
> For `org/mozilla/components`, I think we could use that, I picked `org/mozilla/multiplatform` to separate the components, but there's no need for that, and we can use `org/mozilla/components` if that is easier.
> 
> For `{component}/{version}/{artifact_file_name}` this seems more complicated now. In the example here, we have only one (gradle) component `multiplatform-lib-dummy`, but output the tree from [#9435 (comment)](https://github.com/mozilla-mobile/android-components/pull/9435#issuecomment-762800776) - So kind of a 1:n mapping instead of a 1:1 mapping as before? 🤔

Hm. We may have to assume a default beetmover behavior, used by every other component, and allow for an override for this multiplatform component.

What kind of timeframe+priority are we looking at? Is this just exploratory with no deliverable, or do we have a looming release, or something in between?
